### ISSUE TITLE
Proposal: Type for storing Reactive Context using RxState

### ISSUE THREAD

**Issue by christophbuehler** [rx-angular/rx-angular#1528]
> If there is an interest in this proposal, I will keep it up to date with [the decisions made in regards to Signals](https://github.com/rx-angular/rx-angular/issues/1526). I am also offering to maintain this feature.

# Problem Solved By The Feature

RxAngular has solutions for dealing with [reactive context](https://www.rx-angular.io/docs/template/concepts/reactive-context) within templates:
- [Contextual state for RxIf](https://www.rx-angular.io/docs/template/api/rx-if-directive#inputs)
- [Contextual state for RxLet](https://www.rx-angular.io/docs/template/api/let-directive#inputs)

By default, this reactive context is [inferred from the observable](https://www.rx-angular.io/docs/template/concepts/reactive-context#usage-in-the-template) which is usually not sufficient.
- `Suspense` state can only be the initial state, because [stateful data](https://www.rx-angular.io/docs/state/api/rxjs-operators/stateful#stateful) will never emit `undefined`. So no reload possible.
- Stateful data does not have an error or complete state.

For this reason, RxAngular provides fine-grained control of the reactive context with [triggers](https://www.rx-angular.io/docs/template/api/let-directive#context-trigger). While this provides all the tools needed to work with asynchronous data, working with triggers is challenging and error prone:
- Triggers should not be stateful, because duplicate emissions would be lost (e.g. the api returns the same error twice).
- Triggers are not tied to the data. This makes tasks like passing the context to a child component difficult.
- Naming conventions and boilerplate.

# Solution

I am proposing a type for keeping track of the reactive context using RxState.
|[How that could work (StackBlitz)](https://stackblitz.com/edit/contextful-demo?file=src%2Fexamples%2Ffetch-user.component.ts)|
| --- |

## The State

```typescript
interface RxContextfulState<I, O, E> {
  // The data used to initialize the asynchronous operation.
  request?: I;
  // The last response.
  response?: O;
  // Error thrown by the asynchronous operation.
  error?: E;
  // The context: Suspense, Next, Error or Complete
  context?: RxNotificationKind;
}
...
private readonly state: RxState<{ user: RxContextfulState<string, User> }> = inject(RxState);
```

### fromOperator

Populate `RxContextfulState` from an operator.
[See implementation on StackBlitz](https://stackblitz.com/edit/contextful-demo?file=src%2Fcontextful%2Ffrom-operator.ts).

Returns `RxContextful`:
```typescript
interface RxContextful<I, O, E> {
  request: (value: I) => void;
  setResponse: (value: O) => O;
  request$: Observable<I>;
  response$: Observable<O>;
  error$: Observable<E>;
  context$: Observable<RxNotificationKind>;
  isSuspense$: Observable<boolean>;
  isNext$: Observable<boolean>;
  isError$: Observable<boolean>;
  isComplete$: Observable<boolean>;
}
```
Usage:
```typescript
readonly user = fromOperator<string, User>('user', pipe(
  switchMap((userId) => this.userService.loadUser(userId)),
));
...
<ng-container *rxLet="
  user.response$;
  let user;
  contextTrigger: user.context$
"></ng-container>
```

### fromObservable

```typescript
export function fromObservable<O = undefined, E extends Error = Error>(
  key: string,
  observable: Observable<O>
): Contextful<undefined, O, E> {
  const contextful = fromOperator<undefined, O, E>(key, ($) =>
    $.pipe(switchMap(() => observable))
  );
  contextful.request(void 0);
  return contextful;
}
```

### fromFormControl

Takes care of form control submissions. Also makes reactive forms more reactive by exposing `dirty$` and `valid$`. The scenario for this example: You have a form control for the username and a save button. Display loading, error and success states.

```typescript
readonly saveUsernameContext = fromFormControl<string, Partial<User>>(
  'saveUsername',
  this.usernameControl,
  pipe(switchMap((name) => this.apiService.saveUser({ name })))
);
```

|[From Form Control (StackBlitz)](https://stackblitz.com/edit/contextful-demo?file=src%2Fexamples%2Ffrom-form-control.component.ts)|
| --- |

**Comment by edbzn** [comment]
Hi, first of all, it is a well-written proposal, I like the ideas and the demos! so kudos for that.

I agree with you I think the template reactive context must have its equivalent on the state side, we made something close to your idea outside RxA a few times ago, the [suspensify operator](https://github.com/jscutlery/devkit/tree/main/packages/operators#suspensify). Something like this could definitely have a place in the CDK package.

We could start an RFC directly from this issue to discuss about the scope and the implementation.

**Comment by christophbuehler** [comment]
Hi @edbzn, thanks for your quick and kind reply! I would be really happy to use this issue as an RFC on this topic.

The `suspensify operator` you mentioned strikes me as super readable. Much less is going on behind the curtain and it does not use `RxState` internally. 👏 I will explore this idea further and keep you posted with some demos.

A draft for the scope from my point of view (open for discussion):

## Must-have

- Define a type that contains both data and its reactive context.
- Provide a way of creating an observable of this type from an asynchronous operation.

## Nice-to-have

- Provide a way to connect this state to a form control / form group.

**Comment by hoebbelsB** [comment]
hi @christophbuehler, I fully agree with @edbzn, a very well written proposal :heavy_check_mark: Also, the demos are :man_cook: :kiss: 

I really like the idea of providing the reactive context on a state level. However, I find myself confused with some of your proposed APIs.

### The State

I am getting so confused about the `request` naming, because I just expect it to be an actual `Request`.
As an idea I can throw `payload` | `requestPayload` into the room :)

Also, if possible let's try to have it independent from `RxState` as this increases the flexibility.
We should still think about how we can bridge it / `connect` it into the state, as this can also be useful.

```ts
interface RxContextfulState<I, O, E> {
  // The data used to initialize the asynchronous operation.
  request?: I; // 👈 I am getting so confused about this naming, because I just expect it to be an actual `Request`
  // I would rename this to `payload`
  // The last response.
  response?: O;
  // Error thrown by the asynchronous operation.
  error?: E;
  // The context: Suspense, Next, Error or Complete
  context?: RxNotificationKind;
}

private readonly state: RxState<{ user: RxContextfulState<string, User> }> = inject(RxState);
```

### fromOperator

I like what it does and agree this is super useful, but the way it is used and also named can be improved imho.
I think too much magic is going on behind the scenes. For me it was very unclear what the `key` input is doing until I noticed it is plucking the `user` key from the `RxState` instance to interact with the ContextfulState.
Also the usage of `pipe` is super rarely seen and probably leads to confusion.

I would suggest to simplify the API by removing the `key` input and probably some of the options - but we can go into detail after your next draft :).

But let me quickly showcase what I have in mind right now:

```ts
readonly user = fromOperator<string, User>('user', pipe(
  switchMap((userId) => this.userService.loadUser(userId)),
));

// vs

readonly userWithContext = rxContextful<string>( // 👈 creation function, returning RxContextful
 request$ => request$.pipe( // 👈 expose the `request$` as `Observable`
   switchMap((userId) => this.userService.loadUser(userId)), // 👈 User will be inferred from the return value
 )
)

```

I guess this one `rxContextful` API also makes `fromFormControl` and `fromObservable` kind of obsolete as you can now just switchMap / mergeMap, ... to whatever you like.


### Thought about connecting to RxState

As an addition, you could add a `state$: Observable<RxContextfulState>` property to `RxContextful`, which would let you
easily `connect` it to any state.

```ts

readonly userWithContext = rxContextful<string>(
 request$ => request$.pipe(
   switchMap((userId) => this.userService.loadUser(userId)),
 )
)

constructor() {
  this.state.connect('userContext', this.userWithContext.state$)
}
```

### Nice To Have

Let me also add a nice to have :)

A way to maintain contextual states for entities in a list. If you have a list of `Todo` and display them in a list, each item
basically has it's own contextual information. Users can update single entities, single entities can throw errors, being suspended ...

**Comment by christophbuehler** [comment]
Hi @hoebbelsB, thanks for digging into this!

I am on par with your ideas :)

- `request` is absolutely a confusing name.
- RxState should not be a requirement, so no `key` parameter and no black box.
- As you said, `pipe(` is the uncommon equivalent of `$ => $.pipe(` and should not be used.

---

There are currently two strains of thought on how `Contextful` could be implemented. A quick summary:

## As a factory

```TypeScript
readonly userWithContext = rxContextful<string>( // 👈 returning RxContextful
 request$ => request$.pipe(
   switchMap((userId) => this.userService.loadUser(userId)),
 )
);
```

- 👍 Exposes a function "request? / next? / fetch? / ..." which kicks off the asynchronous operation.
- ❔ No RxState, but should be stateful and handle late subscribers (probably just use [stateful](https://www.rx-angular.io/docs/state/api/rxjs-operators/stateful)?).
- ❔ Maybe connected to state using `state.connect('user', userWithContext.state$)`. An alternative would be to use `rxEffects.register(userWithContext.state$)` taking care of the initial sub, with no need for RxState at all.

## As an operator

```TypeScript
state.connect(
  'user',
  this.loadUser$.pipe(
    switchMap(userId => this.apiService.loadUser(userId)),
    contextful(), // 👈 returning RxContextfulState
  )
);
```

- 👍 It's just an operator, simple and clean. Connecting to state does not need to be explained.
- 👎 A subject à la `loadUser$: Subject<string>()` is often required, causing some boilerplate.

---

If there are no further remarks, I will implement a second draft exploring some more edge cases. 🫡

**Comment by christophbuehler** [comment]
## Draft without RxState

[Try it on StackBlitz](https://stackblitz.com/edit/contextful-demo-cgvznu?file=src%2Fexamples%2Ffetch-user.component.ts)

What I changed based on feedback from @hoebbelsB:
- Removed `key` parameter and usage of `RxState`.
- Hopefully better naming.

```TypeScript
readonly userContext = rxContextful<number, User>(($) =>
  $.pipe(switchMap((userId) => this.apiService.fetchUser(userId)))
);

constructor(private apiService: MockapiService, rxEffects: RxEffects) {
  rxEffects.register(this.userContext.state$);
}
```

I think we cannot infer `User` as suggested. I tried and failed miserably: https://github.com/microsoft/TypeScript/pull/26349

## RFC
What do you think of this?
```TypeScript
readonly userContext = rxContextful<number, User>(($) =>
  $.pipe(switchMap((userId) => this.apiService.fetchUser(userId)))
).register(inject(RxEffects)); // 👈 instead of doing it in the constructor
```

**Comment by edbzn** [comment]
Some thoughts about this, I really prefer the factory and operator way.

> +1 Exposes a function "request? / next? / fetch? / ..." which kicks off the asynchronous operation.

My first point is about the implemented `request` method. I think it is outside the scope of the context. We have the `RxActionsFactory` that can be used to declare UI triggers, having two different ways to trigger an action feels confusing.

My second point is relative to HTTP, it is an open question, do we really want to have something only bound to HTTP, like in the presented implementation where things are relative to request and response or do we prefer something more generic that can be used with every stream?

My last point is about naming, I like `contextful`, but here are some other propositions we can explore: 

- `stateify`
- `suspensify`
- `statusify`
- `observify`
- `contextualize`

In case you need a sandbox app to try out some of your ideas, you can use [this todo app](https://github.com/edbzn/rx-angular-todo-mvc) that implement RxA stuff and consumes an API via HTTP.

**Comment by christophbuehler** [comment]
@edbzn is `RxActionsFactory` part of the public API? Could you please provide a link?

**Comment by christophbuehler** [comment]
> @edbzn: My second point is relative to HTTP, it is an open question, do we really want to have something only bound to HTTP, like in the presented implementation where things are relative to request and response or do we prefer something more generic that can be used with every stream?

The naming of request and response may be a bit irritating. These are not referring to an HTTP request. I think it should be usable with any stream, even without a template.

I like `contextful` or `contextualize`, because they suggest that there is more to context than just a `suspense` state. It is also consistent with the naming of `contextTrigger`.

Based on all the feedback, here are the latest drafts of the two implementations. Looking forward to your inputs!

## As an operator

Using an operator similar to [suspensify](https://github.com/jscutlery/devkit/blob/main/packages/operators/src/lib/suspensify.ts), that takes control of the full lifecycle could look like this:

```TypeScript
readonly ui = inject(RxActionFactory<{ requestUser: number }>).create();

readonly userContext$ = this.ui.requestUser$.pipe(
  rxContextful()( // 👈 currying 🍛 so we can provide a config
    switchMap(userId => this.apiService.fetchUser(userId)),
  ),
);
```

One limitation of this approach is that operators have to return one observable (see what happened to [partition](https://rxjs.dev/api/operators/partition)). So usually some more code is required to get the stateful observables required.

## As a factory

Two suggestions using an optional "initiator" subject. I would not use `RxActionFactory` if this were the implementation of `rxContextful` though. It feels redundant. Any suggestions on how we could make them work together nicely?

```TypeScript
readonly ui = inject(RxActionFactory<{ requestUser: number }>).create();

readonly userContext = rxContextful(($) => $.pipe(
  switchMap((userId) => this.apiService.fetchUser(userId))
), { initiator: this.ui.requestUser$ });

// or

readonly userContext = rxContextful({
  initiator: this.ui.requestUser$,
})(($) => $.pipe(
  switchMap((userId) => this.apiService.fetchUser(userId))),
);
```

**Comment by edbzn** [comment]
Hey @christophbuehler, sorry for the delay. I prefer the operator way, it's more natural than using an `initiator` with a factory. I'm not entirely sure to get why the operator approach requires more code. In your example, it seems to be the opposite. Could you give me your lights on this?

What do others think about the approaches presented above? @rx-angular/core-team

**Comment by christophbuehler** [comment]
Hi @edbzn, thanks for the response. The operator approach leaves us with one observable holding all the state, which is a bit more cumbersome to work with:
`Observable<{ context: RxNotificationKind, requestData: I, data: O, error: E }>`

Besides that I think it's elegant and intuitive. Looking forward to comments from the team.

## A prod example using `fromFormControl`

```TypeScript
readonly descriptionContext = this.fromEditFormControl('description');
```

```HTML
<s1s-standalone-input-host
  [hintText]="t('hint-description')"
  [successText]="t('description-update-success')"
  [context]="descriptionContext"
>
  <tui-input formControlName="description" tuiTextfieldSize="l">
    {{ t('description-label') }}*
  </tui-input>
</s1s-standalone-input-host>
```

What this generic component is capable of: 
- Takes care of dirty and suspense states
- [0:00] Displays a hint
- [0:04] Displays client-side validation error
- [0:10] Successfully save form value
- [0:30] Display server-side error without breaking the chain
- [0:37] Reset form value to last saved state

https://user-images.githubusercontent.com/4547315/228894926-8f770c72-5eec-49f1-a3d2-92b46cc6441b.mov

**Comment by edbzn** [comment]
Old related issue https://github.com/rx-angular/rx-angular/issues/985.

@christophbuehler Did you create a stackblitz implementing your last suggestion?

```ts
readonly userContext$ = this.ui.requestUser$.pipe(
  rxContextful()( // 👈 currying 🍛 so we can provide a config
    switchMap(userId => this.apiService.fetchUser(userId)),
  ),
);
```

I would like to explore it.

**Comment by christophbuehler** [comment]
[Ben Lesh @ State of RxJS April 2023
](https://www.youtube.com/live/MXRPYapMzpo?feature=share&t=1349)

So it is possible to return whatever we want from **rxContextful** if used inside a pipe. So it is definitely worth exploring that idea further. I will work out an example in the next days.

An example "operator function" returning an object instead of an observable:
```TypeScript
const myOperator = (source$: any) => {
  source$.subscribe();
  return {
    state$: EMPTY,
    doSomething: () => 'hi',
    // ...
  };
};
```

**Comment by christophbuehler** [comment]
## New RxJS API

RxJS introduces the new `rx` api (https://github.com/ReactiveX/rxjs/issues/7203#issuecomment-1507834302) in V8 as a "new way of piping". This has been confirmed by Ben Lesh at [State of RxJS April 2023](https://www.youtube.com/live/MXRPYapMzpo?feature=share&t=1340). This function accepts a source observable and a list of `UnaryFunction` instead of `OperatorFunction` (this change also applies to `pipe`!).

```TypeScript
const lorem = rx( // 👈 lorem is of type { obs$: Observable<number>; message: string; }
  obs$, // 👈 first param is converted to an observable
  // 👇 following params are wild west unary functions
  map(() => 3),
  obs$ => ({ obs$, message: 'hi there!' }),
);
```

This opens up new possibilities for `rxContextful`. 🎉

## 1. Improve the "operator" syntax

The operator approach could actually return an object, so [my comment](https://github.com/rx-angular/rx-angular/issues/1528#issuecomment-1460234725) is already outdated. This alleviates the biggest flaw of this API IMHO.

```TypeScript
readonly userContext = this.ui.requestUser$.pipe( // 👈 userContext is an object
  rxContextful( // 👈 let's not do currying here 😿 🍛
    switchMap(userId => this.apiService.fetchUser(userId)),
  ),
);
// 👇 same thing
readonly userContext = rx(
  this.ui.requestUser$,
  rxContextful()(
    switchMap(userId => this.apiService.fetchUser(userId)),
  ),
);
```

## 2. Mimic the `rx` API

This not only addresses the concern of @edbzn of "having two different ways to trigger an action". It would also provide an api that feels more natural to people using the new `rx` syntax:

```TypeScript
rxContextful(
  this.ui.requestUser$,
  switchMap(userId => this.apiService.fetchUser(userId)),
);
```

---

The "operator" syntax still uses a nested callback to be able to recover from errors (is there an alternate way of achieving this)? Did I miss or misunderstand something? Looking forward to your input before implementing the next iteration.

**Comment by hoebbelsB** [comment]
hi @christophbuehler , thank u so much for your efforts here :pray: :clap: .
 
> ## 1. Improve the "operator" syntax

> The operator approach could actually return an object, so [my comment](https://github.com/rx-angular/rx-angular/issues/1528#issuecomment-1460234725) is already outdated. This alleviates the biggest flaw of this API IMHO.
> 
> ```ts
> readonly userContext = this.ui.requestUser$.pipe( // 👈 userContext is an object
>   rxContextful( // 👈 let's not do currying here 😿 🍛
>     switchMap(userId => this.apiService.fetchUser(userId)),
>   ),
> );
> ```

perfect 🤌, I really like that API a lot.

>```ts
> // 👇 same thing
> readonly userContext = rx(
>   this.ui.requestUser$,
>   rxContextful()(
>     switchMap(userId => this.apiService.fetchUser(userId)),
>   ),
> );
> ```

Would it be possible to have the api like the following also for the `rx` operator? It would make it much more readable imo.

```ts

readonly userContext = rx(
 this.ui.requestUser$,
 rxContextful(
   switchMap() // 👈 do not call it as function, but directly pass args
 )
)

```

> ## 2. Mimic the `rx` API

> This not only addresses the concern of @edbzn of "having two different ways to trigger an action". It would also provide an api that feels more natural to people using the new `rx` syntax:
> 
> ```ts
> rxContextful(
>   this.ui.requestUser$,
>   switchMap(userId => this.apiService.fetchUser(userId)),
> );
> ```

also perfect 🤌!

> The "operator" syntax still uses a nested callback to be able to recover from errors (is there an alternate way of achieving this)? Did I miss or misunderstand something? Looking forward to your input before implementing the next iteration.

Can u elaborate on this a bit more? I don't fully understand the problem at this point.

Please keep up the good work, we really want that to be part of the library :pray: Let's move this to a PR soon!

**Comment by christophbuehler** [comment]
Hi @hoebbelsB, firstly, apologies for the late response. I had some other commitments that took precedence, but I'm back on track now. 🚀

I have been using a local version of rxContextful in most of my recent projects and could not imagine working without it. I would love to fast-track this and begin work on a PR asap.

FYI: You can try out the `rx` syntax in [8.0.0-alpha.10](https://www.npmjs.com/package/rxjs/v/8.0.0-alpha.10)!

Yes, we can absolutely pass params to rxContextful directly. This was a mistake on my side.
With nested callbacks I meant the pattern `rx(rxContextful(switchMap()))`, which I don't love.

My issue with the "operator" syntax is that we may lose control over the lifecycle if not used correctly:
```TS
readonly userContext = rx(
  this.ui.requestUser$,
  rxContextful(
    switchMap(userId => this.apiService.fetchUser(userId)),
  ),
  // 👈 everything that follows is wild west
);
```

We don't have that issue with this syntax by design:

```TS
readonly userContext = rxContextful(
  this.ui.requestUser$,
  switchMap(userId => this.apiService.fetchUser(userId)),
  map(user => parseUser(user)), // 👈 throws? We got you covered
);
```

Additionally, there is no need to wait for the `rx` operator to be in a stable release if we go with the second approach.

If it's okay with the team, I would start work on a PR with this. 🙏

**Comment by christophbuehler** [comment]
@hoebbelsB any update on this?

**Comment by edbzn** [comment]
Hey @christophbuehler, sorry for the delay, we discussed this recently and we would like to comeback to it.
